### ISSUE TITLE
[no-unnecessary-type-arguments] fails for CustomEvent

### ISSUE THREAD

**Issue by andreineculau** [typescript-eslint/typescript-eslint#1224]
**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/no-unnecessary-type-arguments": ["error"]
  }
}
```

```TS
class FooEvent extends CustomEvent<any> {
}
```

**Expected Result**

to pass linting.

**Actual Result**

Fails linting, because the type `CustomEvent<T>` has a default type `any` for the generic `T`, and thus it wants just `CustomEvent`.

The problem is that `CustomEvent` is defined as a global variable, thus the meaning of `extends CustomEvent` is different than `extends CustomEvent<any>`.

**Additional Info**

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `2.7.0` |
| `@typescript-eslint/parser`        | `2.7.0` |
| `TypeScript`                       | `3.7.2` |
| `ESLint`                           | `6.6.0` |
| `node`                             | `12.12.0` |
| `npm`                              | `6.11.3` |

**Comment by bradzacher** [comment]
Just use a disable comment for this edge case of global name shadowing.

**Comment by andreineculau** [comment]
Sure but that is not the point of opening an issue, is it? I'm fine with you closing it but why did you close it? Because a workaround exists?

This is not some error on my side nor on typescript's side. If it's not technically possible to detect global name shadowing or it's complex to do so, then at least it would be possible and easy to amend the rule with an option to ignore certain types.

**Comment by bradzacher** [comment]
If I'm understanding your problem correctly, you've got something like:
```ts
// lib.d.ts
class CustomEvent {}

// CustomEvent.ts
class CustomEvent<T = any> {}

// Foo.ts
import { CustomEvent } from './CustomEvent'
class Foo extends CustomEvent {}
class Bar extends CustomEvent<any> {}
```

I have a two questions for you:
- why are you shadowing a global name, instead of naming it something unique?
- [doesn't typescript throw compile time errors at you for defining a class the same a global name?](https://www.typescriptlang.org/play/#code/MYGwhgzhAECyD2BXCBTAogNxQOwC7QG9oBfAKFEhmzAFsVCTzwppMd8iyKXYwAHBl2YwAyig6MA9JOgA6eUA)

IMO if I were a developer on this codebase, I would be very, _very_ confused if I wrote `class Foo extends CustomEvent {}`, and typescript errors because it used the wrong type.
This makes me think it's very much a code smell, that should be fixed at the declaration, instead of fixing the lint rules to allow it.

---

AFAIK it's not possible to detect this shadowing, or if it is - it's very much non-trivial, and not really worth the effort, considering this is a very rare edge-case.

In terms of an ignore option - we avoid them where possible. Ignore options are context-less, which means they over-ignore. I.e. in this case, you only want to ignore this specific class name when extending. But it will also cause the rule to ignore functions/types that have type parameters.

**Comment by andreineculau** [comment]
I can't answer your questions, because I'm not doing what you think I'm doing.

First of all I'm not defining `CustomEvent<T = any>`. It is already defined in `lib.dom.d.ts` since typescript 2.7. See [release notes](https://github.com/microsoft/TypeScript/wiki/Breaking-Changes#typescript-27). So there's nothing to fix declaration-wise.

So my handwritten code is literally just `class FooEvent extends CustomEvent<any> { }`, which fails to lint due to `@typescript-eslint/no-unnecessary-type-arguments`, which has a fixer and rewrites it to `class FooEvent extends CustomEvent { }`, smth that is *very, very* wrong, as you notice as well.

As for the ignore option, I think we're talking about different things. I'm imagining an ignore option that has a list of types i.e. `['CustomEvent']` which will not make `@typescript-eslint/no-unnecessary-type-arguments` yell that `CustomEvent<any>` should be rewritten as `CustomEvent`. The same thing for type parameters as well, even if the two are equivalent.

**Comment by bradzacher** [comment]
> As for the ignore option, I think we're talking about different things

We're talking about the same thing, I'm saying that ignore options are bad because setting `ignore: ['CustomEvent']` will ignore functions, types, variables, classes, etc that are named `CustomEvent`. Which is a lot more far-reaching, and will cause your codebase to be inconsistent.

Eg:
```ts
class Foo extends CustomEvent<any> {} // this should be ignored
const x = new CustomEvent<string>('type', { detail: 'string' }); // this should not be ignored
```

But as mentioned - ignore options are contextless, so they don't know the difference between the two unless you give it context by having one option per usage type: `ignoreClassExtends`, `ignoreNewCall`, etc.
This sounds great, but is a lot of effort to setup in code, and a lot of effort for a user to understand and configure.

---

OOOkay. I am on the same page now. I don't do custom event extensions in TS, so I haven't ever used the type (or know that it existed).
I didn't know that `CustomEvent` was a type was specifically built in to typescript, nor did I know that it is such a complete trainwreck of double naming and weirdness:

```ts
interface CustomEvent<T = any> extends Event {
    /**
     * Returns any custom data event was created with. Typically used for synthetic events.
     */
    readonly detail: T;
    initCustomEvent(typeArg: string, canBubbleArg: boolean, cancelableArg: boolean, detailArg: T): void;
}

declare var CustomEvent: {
    prototype: CustomEvent;
    new<T>(typeArg: string, eventInitDict?: CustomEventInit<T>): CustomEvent<T>;
};
```

Now that I'm on the same page, and understand the context, this looks like a bug in the rule, which is revealed specifically by bad types in typescript.

I say bad types, because:
```ts
interface TInterface<T = any> { }

// breaks

declare const TMyClass: {
    new <T>(): TInterface<T>;
};
class Child extends TMyClass { }
// No base constructor has the specified number of type arguments.

// works

declare const TMyClass2: {
    new <T = any>(): TInterface<T>; // note the default type value
};
class Child2 extends TMyClass2 { }
```

The rule must be using the wrong signature for the pseudo-class.
Now that I know what you're talking about and we have a solid repro, this can be investigated and fixed as the bug it is.

**Comment by bradzacher** [comment]
Minimal repro:
```ts
interface CustomEvent<T = any> {}
declare var CustomEvent: {
  new<T>(): CustomEvent<T>;
};
class Foo extends CustomEvent<T> {}
```

Digging into this a bit, the rule was built from a rule in tslint.
https://github.com/palantir/tslint/blob/master/src/rules/useDefaultTypeParameterRule.ts
Looking at the source code, they both have this bug.

This occurs because the variable has the same name as the backing type.
When the rule interrogates the type of `CustomEvent`, typescript returns an array of declarations. `CustomEvent` has two declarations - an `InterfaceDeclaration`, and a `VariableDeclaration`.
The order of these declarations depends on the source ordering (i.e. in the case of `lib.d.ts`, the `InterfaceDeclaration` appears first in the array).

The code is designed to just use the first compatible declaration, which will always be the interface in this case, as it's the first element in the array. The interface is defined with a default type parameter.

The rule probably needs a "priority" list for the locations. OFC interfaces aren't the correct declaration to use first in the case of an `extends`, but they are in an `implements` clause.

**Comment by andreineculau** [comment]
Cool! Thanks for re-opening and expanding the problem domain, and highlighting the important bits! I think someone (at least me, if I have the time) has enough context to file a PR on this.

> Which is a lot more far-reaching, and will cause your codebase to be inconsistent.

True, but having a naïve "ignore" option is the equivalent of very low inconsistency as opposed to 
* high inconsistency with this rule as active, but buggy linter and broken code (due to the fixer)
* no consistency with this rule as inactive.

I'm saying this given that I only know of CustomEvent, my TypeScript colleagues didn't know about it, you didn't know about it, so the guess would be that it is indeed a corner case, rather than an epidemic (which would make the "ignore" option irrelevant to begin with).

In any way, your reasoning is far more appropriate than having an ignore option, given that it's feasible to implement.

**Comment by armano2** [comment]
small note:
CustomEvent is not typescript specific thing, this is part of web api, 
from context of browser this is a class

https://developer.mozilla.org/en-US/docs/Web/API/CustomEvent
https://developer.mozilla.org/en-US/docs/Web/Guide/Events/Creating_and_triggering_events

**Comment by Himself65** [comment]
same issue here
